### ISSUE TITLE
Hover state doesn't update if there are no mouse move events

### ISSUE THREAD

**Issue by nimble0** [mikke89/RmlUi#220]
Reproducing:

- Load the animation sample.
- Position the mouse over the "Events (press arrow keys)" box. The box will become highlighted from the hover style.
- Don't move the mouse.
- Press arrow keys to move the box.

Result: The box will remain highlighted even when the cursor is no longer over the box.

It can cause some less innocuous bugs too; I encountered it while trying to temporarily block input in a key mapping screen.

A simple fix is to make sure to call ProcessMouseMove before every update (the arguments can be the same as last update).

**Comment by mikke89** [comment]
Right, yes, this also happens with animations if the hovered element moves away from the mouse cursor. Another case is when the element under the cursor is removed, and new ones are constructed. I've noticed this may break scrolling until after the mouse has moved again, probably because we need to re-establish a new hover element. This is noticeable sometimes in the debugger.

I'd like to know more about your issue with your key mapping screen and how that is affected, could you add some more details?

I'm concerned the proposed fix may cause some unexpected behavior, in particular for users not using mouse cursors or intentionally not providing any mouse input. Hm, actually, now after looking over the code, I see that we don't send a mouse move event if the position is the same, so this might be a viable approach. Although we do send out drag events, not sure if this is an issue.

In any case we should think some more about possible ways this might cause issues before we push these changes.

**Comment by nimble0** [comment]
Oh, I meant that as a user side fix.

> I'd like to know more about your issue with your key mapping screen and how that is affected, could you add some more details?

I have a list of buttons corresponding to actions that can be mapped. To map an action you click the button for that action and the page starts waiting for an input (key press or mouse button press). I don't want the input to do whatever action it would normally do so I use "pointer-events: none" to block them while waiting for input. This works fine unless the user doesn't move the mouse and then presses the left mouse button. In this case it still registers the click on the button despite "pointer-events: none" and starts waiting for input again.

**Comment by mikke89** [comment]
Ah, I see. It could be a place to start in any case.

Thanks for your very descriptive example.

One idea I have for a solution is to do something equivalent to `ProcessMouseMove` during `Context::Update`. However, we also add a new function `Context::ProcessMouseOut` or similar which disables this automatic processing to avoid unintended (de)focusing. The automatic processing resumes after the user calls `Context::ProcessMouseMove`.

Any thoughts or other ideas?

**Comment by nimble0** [comment]
Sounds good to me.

**Comment by xland** [comment]
@mikke89 

How about this issue's progress?
It's really important to me.
Is it possible to fix it in 4.x?
Thanks very much.

**Comment by mikke89** [comment]
Yeah, it is planned for RmlUi 5.0, I'll try to prioritize this in the near feature.

I won't backport any new features, if you really need it in 4.x you will have to do that yourself. There aren't really any significant breaking changes (at least yet), so you might want to consider using the RmlUi master branch for your project.

**Comment by xland** [comment]
Thanks for your reply.
I hope to see it resolved on the master branch soon.

**Comment by mikke89** [comment]
Hey.

I've made an effort at implementing this now.

- The hover state of any elements under the mouse will now automatically be updated during `Context::Update()`.
- Added `Context::ProcessMouseLeave()` which ensures that the hovered state is removed from all elements, and stops the context update from automatically hovering elements.
- When `Context::ProcessMouseMove()` is called next the context update will start updating hover states again.
- I've also updated all the backends to handle mouse leave events.

Seems to work well for me for now, however it does require more testing in case of unintended consequences. Let me know how it goes, cheers!
